CMake fails to detect NEON, etc. support on ARM processor

##### System information (version)

- OpenCV => 4.4.0
- Operating System / Platform => Ubuntu 18.04 / armv7l
- Compiler => gcc 7.5.0

##### Detailed description

CMake does not detect that my ARM CPU supports NEON instructions or other features. If I compile the example source code from `cmake/checks/cpu_neon.cpp` with `-mfpu=neon`, it builds and runs successfully.

```
-- Performing Test HAVE_CXX_MFPU_VFPV3
-- Performing Test HAVE_CXX_MFPU_VFPV3 - Success
-- Performing Test HAVE_CPU_NEON_SUPPORT (check file: cmake/checks/cpu_neon.cpp)
-- Performing Test HAVE_CPU_NEON_SUPPORT - Failed
-- Performing Test HAVE_CXX_MFPU_NEON (check file: cmake/checks/cpu_neon.cpp)
-- Performing Test HAVE_CXX_MFPU_NEON - Success
-- Performing Test HAVE_CPU_FP16_SUPPORT (check file: cmake/checks/cpu_fp16.cpp)
-- Performing Test HAVE_CPU_FP16_SUPPORT - Failed
-- Performing Test HAVE_CXX_MFPU_NEON_FP16_MFP16_FORMAT_IEEE (check file: cmake/checks/cpu_fp16.cpp)
-- Performing Test HAVE_CXX_MFPU_NEON_FP16_MFP16_FORMAT_IEEE - Success
-- Performing Test HAVE_CPU_BASELINE_FLAGS
-- Performing Test HAVE_CPU_BASELINE_FLAGS - Success

...

--   CPU/HW features:
--     Baseline:
--       requested:                 DETECT
--       disabled:                  VFPV3 NEON
```

* [CMakeCache.txt](https://github.com/opencv/opencv/files/5159917/CMakeCache.txt)
* [CMakeError.log](https://github.com/opencv/opencv/files/5159918/CMakeError.log)
* [CMakeOutput.log](https://github.com/opencv/opencv/files/5159919/CMakeOutput.log)

##### Steps to reproduce

```
cmake .. \
    -DCMAKE_BUILD_TYPE=RelWithDebugInfo \
    -DCMAKE_INSTALL_PREFIX="$(cd root; pwd)" \
    -DOPENCV_EXTRA_MODULES_PATH=../opencv_contrib/modules \
    -DBUILD_JAVA=OFF \
    -DBUILD_TESTS=OFF \
    -DBUILD_PERF_TESTS=OFF \
    -DBUILD_opencv_java=OFF \
    -DBUILD_opencv_python2=OFF \
    -DBUILD_opencv_python3=ON \
    -DPYTHON_DEFAULT_EXECUTABLE="$(command -v python3.7)" \
    -DPYTHON3_INCLUDE_DIR="$PYTHON_INCLUDE" \
    -DPYTHON3_LIBRARY="$PYTHON_LIBRARY" \
    -DPYTHON3_EXECUTABLE="$(command -v python3.7)" \
    -DPYTHON3_NUMPY_INCLUDE_DIRS="$NUMPY_INCLUDE" \
    -DPYTHON3_PACKAGES_PATH="$PYTHON_PACKAGES" \
    -DOPENCV_SKIP_PYTHON_LOADER=ON
```

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   <!--
   OpenCV team works with answers.opencv.org, Stack Overflow and other communities
   to discuss problems. Tickets with question without real issue statement will be
   closed.
   -->
 - [x] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
   <!--
   Places to check:
   * OpenCV documentation: https://docs.opencv.org
   * FAQ page: https://github.com/opencv/opencv/wiki/FAQ
   * OpenCV forum: https://answers.opencv.org
   * OpenCV issue tracker: https://github.com/opencv/opencv/issues?q=is%3Aissue
   * Stack Overflow branch: https://stackoverflow.com/questions/tagged/opencv
   -->
 - [x] I updated to latest OpenCV version and the issue is still there
   <!--
   master branch for OpenCV 4.x and 3.4 branch for OpenCV 3.x releases.
   OpenCV team supports only latest release for each branch.
   The ticket is closed, if the problem is not reproduced with modern version.
   -->
 - [x] There is reproducer code and related data files: videos, images, onnx, etc
   <!--
   The best reproducer -- test case for OpenCV that we can add to the library.
   Recommendations for media files and binary files:
   * Try to reproduce the issue with images and videos in opencv_extra repository
     to reduce attachment size
   * Use PNG for images, if you report some CV related bug, but not image reader
     issue
   * Attach the image as archite to the ticket, if you report some reader issue.
     Image hosting services compress images and it breaks the repro code.
   * Provide ONNX file for some public model or ONNX file with with random weights,
     if you report ONNX parsing or handling issue. Architecture details diagram
     from netron tool can be very useful too. See https://lutzroeder.github.io/netron/
   -->

```
Building CXX object CMakeFiles/cmTC_13884.dir/cpu_neon.cpp.o
/usr/bin/c++    -fPIE   -std=c++11 -o CMakeFiles/cmTC_13884.dir/cpu_neon.cpp.o -c /home/ubuntu/opencv-4.4.0/cmake/checks/cpu_neon.cpp
/home/ubuntu/opencv-4.4.0/cmake/checks/cpu_neon.cpp:24:2: error: #error "NEON is not supported"
```

It doesn't pass *any* argument here to test whether the compiler supports NEON. It needs to pass `-mcpu=neon` to get `__ARM_NEON__` defined.

I am not sure why `HAVE_CXX_MFPU_NEON_FP16_MFP16_FORMAT_IEEE` succeeded when the others failed.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -486,7 +486,7 @@ OCV_OPTION(OPENCV_ENABLE_MEMORY_SANITIZER "Better support for memory/address san
 OCV_OPTION(ENABLE_OMIT_FRAME_POINTER  "Enable -fomit-frame-pointer for GCC"                      ON   IF CV_GCC )
 OCV_OPTION(ENABLE_POWERPC             "Enable PowerPC for GCC"                                   ON   IF (CV_GCC AND CMAKE_SYSTEM_PROCESSOR MATCHES powerpc.*) )
 OCV_OPTION(ENABLE_FAST_MATH           "Enable compiler options for fast math optimizations on FP computations (not recommended)" OFF)
-if(NOT IOS)  # Use CPU_BASELINE instead
+if(NOT IOS AND CMAKE_CROSSCOMPILING)  # Use CPU_BASELINE instead
 OCV_OPTION(ENABLE_NEON                "Enable NEON instructions"                                 (NEON OR ANDROID_ARM_NEON OR AARCH64) IF (CV_GCC OR CV_CLANG) AND (ARM OR AARCH64 OR IOS) )
 OCV_OPTION(ENABLE_VFPV3               "Enable VFPv3-D32 instructions"                            OFF  IF (CV_GCC OR CV_CLANG) AND (ARM OR AARCH64 OR IOS) )
 endif()

